File: cpp/src/provider/hltb.cpp
#include <retro_metadata/cache/cache.hpp>
#include <retro_metadata/config.hpp>
#include <retro_metadata/errors.hpp>
#include <retro_metadata/internal/matching.hpp>
#include <retro_metadata/provider/base.hpp>
#include <retro_metadata/provider/provider.hpp>
#include <retro_metadata/provider/registry.hpp>

#include <cpr/cpr.h>
#include <nlohmann/json.hpp>

#include <regex>

namespace retro_metadata {

namespace {

const std::string kHltbImageUrl = "https://howlongtobeat.com/games/";
const std::string kGithubHltbApiUrl =
    "https://raw.githubusercontent.com/rommapp/romm/refs/heads/master/backend/handler/metadata/"
    "fixtures/hltb_api_url";
const std::string kDefaultSearchEndpoint = "search";

const std::regex kHltbTagRegex(R"(\(hltb-(\d+)\))", std::regex::icase);

// Helper to safely get string from JSON
std::string get_string(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_string()) {
        return j[key].get<std::string>();
    }
    return "";
}

// Helper to safely get number from JSON
double get_number(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_number()) {
        return j[key].get<double>();
    }
    return 0.0;
}

// Clean filename for search
std::string clean_filename(const std::string& filename) {
    static const std::regex ext_pattern(R"(\.[^.]+$)");
    static const std::regex tag_pattern(R"(\s*[\(\[][^\)\]]*[\)\]])");

    std::string name = std::regex_replace(filename, ext_pattern, "");
    name = std::regex_replace(name, tag_pattern, "");

    // Trim
    auto start = name.find_first_not_of(" \t\n\r");
    if (start == std::string::npos) return "";
    auto end = name.find_last_not_of(" \t\n\r");
    return name.substr(start, end - start + 1);
}

}  // namespace

/// @brief HowLongToBeat metadata provider
class HLTBProvider : public Provider {
public:
    HLTBProvider(const ProviderConfig& config, std::shared_ptr<Cache> cache)
        : config_(config), cache_(std::move(cache)) {}

    std::string name() const override { return "hltb"; }

    std::vector<SearchResult> search(const std::string& query, const SearchOptions& opts) override {
        if (!config_.enabled) {
            return {};
        }

        int limit = opts.limit > 0 ? opts.limit : 20;
        auto result = request("search", build_search_data(query, limit));

        if (!result.contains("data") || !result["data"].is_array()) {
            return {};
        }

        std::vector<SearchResult> results;
        for (const auto& item : result["data"]) {
            if (!item.is_object()) continue;

            int game_id = static_cast<int>(get_number(item, "game_id"));
            if (game_id == 0) continue;

            SearchResult sr;
            sr.name = get_string(item, "game_name");
            sr.provider = name();
            sr.provider_id = game_id;

            std::string img = get_string(item, "game_image");
            if (!img.empty()) {
                sr.cover_url = kHltbImageUrl + img;
            }

            int year = static_cast<int>(get_number(item, "release_world"));
            if (year > 0) {
                sr.release_year = year;
            }

            std::string platform = get_string(item, "profile_platform");
            if (!platform.empty()) {
                // Split by comma
                size_t pos = 0;
                while (pos < platform.size()) {
                    size_t comma = platform.find(", ", pos);
                    if (comma == std::string::npos) {
                        sr.platforms.push_back(platform.substr(pos));
                        break;
                    }
                    sr.platforms.push_back(platform.substr(pos, comma - pos));
                    pos = comma + 2;
                }
            }

            results.push_back(std::move(sr));
        }

        return results;
    }

    std::unique_ptr<GameResult> get_by_id(int game_id) override {
        if (!config_.enabled) {
            return nullptr;
        }

        auto search_data = build_search_data("", 1);
        search_data["gameId"] = game_id;

        auto result = request("search", search_data);

        if (!result.contains("data") || !result["data"].is_array() ||
            result["data"].empty()) {
            return nullptr;
        }

        return build_game_result(result["data"][0]);
    }

    std::unique_ptr<GameResult> identify(
        const std::string& filename, const IdentifyOptions& opts) override {
        if (!config_.enabled) {
            return nullptr;
        }

        // Check for HLTB ID tag in filename
        std::smatch match;
        if (std::regex_search(filename, match, kHltbTagRegex) && match.size() > 1) {
            try {
                int tagged_id = std::stoi(match[1].str());
                auto result = get_by_id(tagged_id);
                if (result) {
                    return result;
                }
            } catch (...) {
                // Ignore parse errors
            }
        }

        // Clean the filename and search
        std::string search_term = clean_filename(filename);
        auto result = request("search", build_search_data(search_term, 20));

        if (!result.contains("data") || !result["data"].is_array() ||
            result["data"].empty()) {
            return nullptr;
        }

        // Build name to game map
        std::map<std::string, nlohmann::json> games_by_name;
        std::vector<std::string> names;
        for (const auto& item : result["data"]) {
            std::string game_name = get_string(item, "game_name");
            if (!game_name.empty()) {
                games_by_name[game_name] = item;
                names.push_back(game_name);
            }
        }

        // Find best match
        auto best = matching::find_best_match_simple(search_term, names);
        if (!best.found()) {
            return nullptr;
        }

        auto game_result = build_game_result(games_by_name[best.match]);
        game_result->match_score = best.score;
        return game_result;
    }

    void heartbeat() override {
        if (!config_.enabled) {
            throw ConnectionError(name(), "provider is disabled");
        }

        // Try to fetch the security token to check connectivity
        std::string token = fetch_security_token();
        if (token.empty()) {
            throw ConnectionError(name(), "failed to get security token");
        }
    }

    void close() override {
        // No resources to clean up
    }

private:
    nlohmann::json build_search_data(const std::string& query, int limit) {
        nlohmann::json search_terms = nlohmann::json::array();
        if (!query.empty()) {
            // Split by space
            std::istringstream iss(query);
            std::string word;
            while (iss >> word) {
                search_terms.push_back(word);
            }
        }

        return {
            {"searchType", "games"},
            {"searchTerms", search_terms},
            {"searchPage", 1},
            {"size", limit},
            {"searchOptions",
             {{"games",
               {{"userId", 0},
                {"platform", ""},
                {"sortCategory", "popular"},
                {"rangeCategory", "main"},
                {"rangeTime", {{"min", 0}, {"max", 0}}},
                {"gameplay", {{"perspective", ""}, {"flow", ""}, {"genre", ""}}},
                {"modifier", ""}}},
              {"users", {{"sortCategory", "postcount"}}},
              {"filter", ""},
              {"sort", 0},
              {"randomizer", 0}}}};
    }

    std::string fetch_search_endpoint() {
        if (!search_endpoint_.empty()) {
            return search_endpoint_;
        }

        cpr::Response r = cpr::Get(cpr::Url{kGithubHltbApiUrl});
        if (r.status_code == 200 && !r.text.empty()) {
            // Trim
            search_endpoint_ = r.text;
            auto start = search_endpoint_.find_first_not_of(" \t\n\r");
            auto end = search_endpoint_.find_last_not_of(" \t\n\r");
            if (start != std::string::npos) {
                search_endpoint_ = search_endpoint_.substr(start, end - start + 1);
            }
        }

        if (search_endpoint_.empty()) {
            search_endpoint_ = kDefaultSearchEndpoint;
        }

        return search_endpoint_;
    }

    std::string fetch_security_token() {
        if (!security_token_.empty()) {
            return security_token_;
        }

        cpr::Response r = cpr::Get(
            cpr::Url{base_url_ + "/search/init"},
            cpr::Header{{"User-Agent", user_agent_}});

        if (r.status_code == 200) {
            try {
                auto json = nlohmann::json::parse(r.text);
                if (json.contains("token") && json["token"].is_string()) {
                    security_token_ = json["token"].get<std::string>();
                }
            } catch (...) {
                // Ignore parse errors
            }
        }

        return security_token_;
    }

    nlohmann::json request(const std::string& endpoint, const nlohmann::json& data) {
        std::string actual_endpoint = endpoint;
        if (endpoint == "search") {
            actual_endpoint = fetch_search_endpoint();
        }

        std::string url = base_url_ + "/" + actual_endpoint;
        std::string body = data.dump();

        cpr::Header headers = {
            {"User-Agent", user_agent_},
            {"Content-Type", "application/json"},
            {"Origin", "https://howlongtobeat.com"},
            {"Referer", "https://howlongtobeat.com"}};

        std::string token = fetch_security_token();
        if (!token.empty()) {
            headers["X-Auth-Token"] = token;
        }

        cpr::Response r = cpr::Post(cpr::Url{url}, headers, cpr::Body{body});

        if (r.status_code != 200) {
            throw ConnectionError(name(), "HTTP " + std::to_string(r.status_code));
        }

        try {
            return nlohmann::json::parse(r.text);
        } catch (const std::exception& e) {
            throw ConnectionError(name(), std::string("JSON parse error: ") + e.what());
        }
    }

    std::unique_ptr<GameResult> build_game_result(const nlohmann::json& game) {
        auto result = std::make_unique<GameResult>();

        int game_id = static_cast<int>(get_number(game, "game_id"));
        result->name = get_string(game, "game_name");
        result->provider = name();
        result->provider_id = game_id;
        result->provider_ids = {{"hltb", game_id}};

        std::string img = get_string(game, "game_image");
        if (!img.empty()) {
            result->artwork.cover_url = kHltbImageUrl + img;
        }

        // Metadata
        int year = static_cast<int>(get_number(game, "release_world"));
        if (year > 0) {
            result->metadata.release_year = year;
        }

        double score = get_number(game, "review_score");
        if (score > 0) {
            result->metadata.total_rating = score;
        }

        if (get_number(game, "comp_main") > 0) {
            result->metadata.game_modes.push_back("Single Player");
        }
        if (get_number(game, "comp_plus") > 0) {
            result->metadata.game_modes.push_back("Completionist");
        }

        result->metadata.developer = get_string(game, "profile_dev");

        // Raw data
        result->metadata.raw_data = {
            {"main_story", get_number(game, "comp_main")},
            {"main_plus_extras", get_number(game, "comp_plus")},
            {"completionist", get_number(game, "comp_100")},
            {"all_styles", get_number(game, "comp_all")},
            {"profile_popular", get_number(game, "profile_popular")},
            {"count_comp", get_number(game, "count_comp")},
            {"count_playing", get_number(game, "count_playing")},
            {"count_backlog", get_number(game, "count_backlog")},
            {"count_replay", get_number(game, "count_replay")},
            {"count_retired", get_number(game, "count_retired")},
            {"review_score", get_number(game, "review_score")}};

        result->raw_response = game;

        return result;
    }

    ProviderConfig config_;
    std::shared_ptr<Cache> cache_;
    std::string base_url_ = "https://howlongtobeat.com/api";
    std::string user_agent_ = "retro-metadata/1.0";
    std::string search_endpoint_;
    std::string security_token_;
};

// Auto-register the provider
namespace {
[[maybe_unused]] ProviderRegistrar hltb_registrar(
    "hltb",
    [](const ProviderConfig& config, std::shared_ptr<Cache> cache) {
        return std::make_unique<HLTBProvider>(config, std::move(cache));
    });
}  // namespace

}  // namespace retro_metadata

File: pkg/provider/hltb/hltb.go
// Package hltb provides metadata from HowLongToBeat.
package hltb

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"

	retrometadata "github.com/josegonzalez/retro-metadata/pkg/retrometadata"
)

const (
	hltbImageURL          = "https://howlongtobeat.com/games/"
	githubHLTBAPIURL      = "https://raw.githubusercontent.com/rommapp/romm/refs/heads/master/backend/handler/metadata/fixtures/hltb_api_url"
	defaultSearchEndpoint = "search"
)

var (
	// Regex to detect HLTB ID tags in filenames like (hltb-12345)
	hltbTagRegex = regexp.MustCompile(`(?i)\(hltb-(\d+)\)`)

	// ErrProviderDisabled is returned when the provider is disabled.
	ErrProviderDisabled = fmt.Errorf("provider is disabled")
)

// Provider implements the HowLongToBeat metadata provider.
type Provider struct {
	config         *retrometadata.ProviderConfig
	client         *http.Client
	baseURL        string
	userAgent      string
	securityToken  string
	searchEndpoint string
}

// New creates a new HLTB provider.
func New(config *retrometadata.ProviderConfig) *Provider {
	timeout := time.Duration(config.Timeout) * time.Second
	if timeout == 0 {
		timeout = 30 * time.Second
	}

	return &Provider{
		config:    config,
		client:    &http.Client{Timeout: timeout},
		baseURL:   "https://howlongtobeat.com/api",
		userAgent: "retro-metadata/1.0",
	}
}

// Name returns the provider name.
func (p *Provider) Name() string {
	return "hltb"
}

func (p *Provider) fetchSearchEndpoint(ctx context.Context) string {
	if p.searchEndpoint != "" {
		return p.searchEndpoint
	}

	req, err := http.NewRequestWithContext(ctx, "GET", githubHLTBAPIURL, nil)
	if err != nil {
		p.searchEndpoint = defaultSearchEndpoint
		return p.searchEndpoint
	}

	resp, err := p.client.Do(req)
	if err != nil {
		p.searchEndpoint = defaultSearchEndpoint
		return p.searchEndpoint
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		p.searchEndpoint = defaultSearchEndpoint
		return p.searchEndpoint
	}

	p.searchEndpoint = strings.TrimSpace(string(body))
	return p.searchEndpoint
}

func (p *Provider) fetchSecurityToken(ctx context.Context) string {
	if p.securityToken != "" {
		return p.securityToken
	}

	req, err := http.NewRequestWithContext(ctx, "GET", p.baseURL+"/search/init", nil)
	if err != nil {
		return ""
	}

	req.Header.Set("User-Agent", p.userAgent)

	resp, err := p.client.Do(req)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return ""
	}

	if token, ok := result["token"].(string); ok {
		p.securityToken = token
	}

	return p.securityToken
}

func (p *Provider) request(ctx context.Context, endpoint string, data map[string]interface{}) (map[string]interface{}, error) {
	// Use dynamic search endpoint if this is a search request
	if endpoint == "search" {
		endpoint = p.fetchSearchEndpoint(ctx)
	}

	url := p.baseURL + "/" + endpoint

	jsonData, err := json.Marshal(data)
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(jsonData))
	if err != nil {
		return nil, err
	}

	req.Header.Set("User-Agent", p.userAgent)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Origin", "https://howlongtobeat.com")
	req.Header.Set("Referer", "https://howlongtobeat.com")

	// Add security token if available
	if token := p.fetchSecurityToken(ctx); token != "" {
		req.Header.Set("X-Auth-Token", token)
	}

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, &retrometadata.ConnectionError{Provider: p.Name(), Details: err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, &retrometadata.ConnectionError{Provider: p.Name(), Details: fmt.Sprintf("HTTP %d", resp.StatusCode)}
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result, nil
}

func buildSearchData(query string, limit int) map[string]interface{} {
	return map[string]interface{}{
		"searchType":  "games",
		"searchTerms": strings.Split(query, " "),
		"searchPage":  1,
		"size":        limit,
		"searchOptions": map[string]interface{}{
			"games": map[string]interface{}{
				"userId":        0,
				"platform":      "",
				"sortCategory":  "popular",
				"rangeCategory": "main",
				"rangeTime":     map[string]interface{}{"min": 0, "max": 0},
				"gameplay":      map[string]interface{}{"perspective": "", "flow": "", "genre": ""},
				"modifier":      "",
			},
			"users":      map[string]interface{}{"sortCategory": "postcount"},
			"filter":     "",
			"sort":       0,
			"randomizer": 0,
		},
	}
}

// Search searches for games by name.
func (p *Provider) Search(ctx context.Context, query string, opts retrometadata.SearchOptions) ([]retrometadata.SearchResult, error) {
	if !p.config.Enabled {
		return nil, nil
	}

	limit := opts.Limit
	if limit == 0 {
		limit = 20
	}

	result, err := p.request(ctx, "search", buildSearchData(query, limit))
	if err != nil {
		return nil, err
	}

	data, ok := result["data"].([]interface{})
	if !ok {
		return nil, nil
	}

	var results []retrometadata.SearchResult
	for _, item := range data {
		game, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		gameID := int(getFloat64(game, "game_id"))
		if gameID == 0 {
			continue
		}

		coverURL := ""
		if img := getString(game, "game_image"); img != "" {
			coverURL = hltbImageURL + img
		}

		var releaseYear *int
		if year := int(getFloat64(game, "release_world")); year > 0 {
			releaseYear = &year
		}

		platforms := []string{}
		if platform := getString(game, "profile_platform"); platform != "" {
			platforms = strings.Split(platform, ", ")
		}

		results = append(results, retrometadata.SearchResult{
			Name:        getString(game, "game_name"),
			Provider:    p.Name(),
			ProviderID:  gameID,
			CoverURL:    coverURL,
			Platforms:   platforms,
			ReleaseYear: releaseYear,
		})
	}

	return results, nil
}

// GetByID gets game details by HLTB ID.
func (p *Provider) GetByID(ctx context.Context, gameID int) (*retrometadata.GameResult, error) {
	if !p.config.Enabled {
		return nil, nil
	}

	searchData := buildSearchData("", 1)
	searchData["gameId"] = gameID

	result, err := p.request(ctx, "search", searchData)
	if err != nil {
		return nil, err
	}

	data, ok := result["data"].([]interface{})
	if !ok || len(data) == 0 {
		return nil, nil
	}

	game, ok := data[0].(map[string]interface{})
	if !ok {
		return nil, nil
	}

	return p.buildGameResult(game), nil
}

// Identify identifies a game from a ROM filename.
func (p *Provider) Identify(ctx context.Context, filename string, opts retrometadata.IdentifyOptions) (*retrometadata.GameResult, error) {
	if !p.config.Enabled {
		return nil, nil
	}

	// Check for HLTB ID tag in filename
	if matches := hltbTagRegex.FindStringSubmatch(filename); len(matches) > 1 {
		var taggedID int
		if _, err := fmt.Sscanf(matches[1], "%d", &taggedID); err == nil {
			result, err := p.GetByID(ctx, taggedID)
			if err == nil && result != nil {
				return result, nil
			}
		}
	}

	// Clean the filename
	searchTerm := cleanFilename(filename)

	// Search for the game
	result, err := p.request(ctx, "search", buildSearchData(searchTerm, 20))
	if err != nil {
		return nil, err
	}

	data, ok := result["data"].([]interface{})
	if !ok || len(data) == 0 {
		return nil, nil
	}

	// Build name to game map
	gamesByName := make(map[string]map[string]interface{})
	for _, item := range data {
		if game, ok := item.(map[string]interface{}); ok {
			if name := getString(game, "game_name"); name != "" {
				gamesByName[name] = game
			}
		}
	}

	// Find best match
	var names []string
	for name := range gamesByName {
		names = append(names, name)
	}

	bestMatch, score := findBestMatch(searchTerm, names)
	if bestMatch == "" {
		return nil, nil
	}

	game := gamesByName[bestMatch]
	gameResult := p.buildGameResult(game)
	gameResult.MatchScore = score
	return gameResult, nil
}

func (p *Provider) buildGameResult(game map[string]interface{}) *retrometadata.GameResult {
	gameID := int(getFloat64(game, "game_id"))

	coverURL := ""
	if img := getString(game, "game_image"); img != "" {
		coverURL = hltbImageURL + img
	}

	metadata := p.extractMetadata(game)

	providerID := gameID
	return &retrometadata.GameResult{
		Name:       getString(game, "game_name"),
		Provider:   p.Name(),
		ProviderID: &providerID,
		ProviderIDs: map[string]int{
			"hltb": gameID,
		},
		Artwork: retrometadata.Artwork{
			CoverURL: coverURL,
		},
		Metadata:    metadata,
		RawResponse: game,
	}
}

func (p *Provider) extractMetadata(game map[string]interface{}) retrometadata.GameMetadata {
	var releaseYear *int
	if year := int(getFloat64(game, "release_world")); year > 0 {
		releaseYear = &year
	}

	var totalRating *float64
	if score := getFloat64(game, "review_score"); score > 0 {
		totalRating = &score
	}

	gameModes := []string{}
	if getFloat64(game, "comp_main") > 0 {
		gameModes = append(gameModes, "Single Player")
	}
	if getFloat64(game, "comp_plus") > 0 {
		gameModes = append(gameModes, "Completionist")
	}

	developer := getString(game, "profile_dev")

	platformsStr := getString(game, "profile_platform")
	platforms := []string{}
	if platformsStr != "" {
		platforms = strings.Split(platformsStr, ",")
		for i := range platforms {
			platforms[i] = strings.TrimSpace(platforms[i])
		}
	}

	return retrometadata.GameMetadata{
		ReleaseYear: releaseYear,
		TotalRating: totalRating,
		GameModes:   gameModes,
		Developer:   developer,
		RawData: map[string]any{
			"main_story":       getFloat64(game, "comp_main"),
			"main_plus_extras": getFloat64(game, "comp_plus"),
			"completionist":    getFloat64(game, "comp_100"),
			"all_styles":       getFloat64(game, "comp_all"),
			"platforms":        platforms,
			"profile_popular":  getFloat64(game, "profile_popular"),
			"count_comp":       getFloat64(game, "count_comp"),
			"count_playing":    getFloat64(game, "count_playing"),
			"count_backlog":    getFloat64(game, "count_backlog"),
			"count_replay":     getFloat64(game, "count_replay"),
			"count_retired":    getFloat64(game, "count_retired"),
			"review_score":     getFloat64(game, "review_score"),
		},
	}
}

// GetCompletionTimes returns completion times for a game.
func (p *Provider) GetCompletionTimes(ctx context.Context, gameID int) (map[string]float64, error) {
	result, err := p.GetByID(ctx, gameID)
	if err != nil || result == nil {
		return nil, err
	}

	times := make(map[string]float64)
	if rawData := result.Metadata.RawData; rawData != nil {
		if v, ok := rawData["main_story"].(float64); ok {
			times["main_story"] = v
		}
		if v, ok := rawData["main_plus_extras"].(float64); ok {
			times["main_plus_extras"] = v
		}
		if v, ok := rawData["completionist"].(float64); ok {
			times["completionist"] = v
		}
		if v, ok := rawData["all_styles"].(float64); ok {
			times["all_styles"] = v
		}
	}

	return times, nil
}

// Heartbeat checks if the provider is available.
func (p *Provider) Heartbeat(ctx context.Context) error {
	if !p.config.Enabled {
		return ErrProviderDisabled
	}

	// Try to fetch the security token to check connectivity
	token := p.fetchSecurityToken(ctx)
	if token == "" {
		return &retrometadata.ConnectionError{Provider: p.Name(), Details: "failed to get security token"}
	}
	return nil
}

// Close closes the provider.
func (p *Provider) Close() error {
	return nil
}

// Helper functions

func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

func getFloat64(m map[string]interface{}, key string) float64 {
	if v, ok := m[key].(float64); ok {
		return v
	}
	return 0
}

func cleanFilename(filename string) string {
	name := regexp.MustCompile(`\.[^.]+$`).ReplaceAllString(filename, "")
	name = regexp.MustCompile(`\s*[\(\[][^\)\]]*[\)\]]`).ReplaceAllString(name, "")
	return strings.TrimSpace(name)
}

func findBestMatch(query string, candidates []string) (string, float64) {
	if len(candidates) == 0 {
		return "", 0
	}

	queryLower := strings.ToLower(query)
	bestMatch := ""
	bestScore := 0.0

	for _, candidate := range candidates {
		candidateLower := strings.ToLower(candidate)

		if candidateLower == queryLower {
			return candidate, 1.0
		}

		if strings.Contains(candidateLower, queryLower) || strings.Contains(queryLower, candidateLower) {
			score := float64(len(queryLower)) / float64(max(len(queryLower), len(candidateLower)))
			if score > bestScore {
				bestScore = score
				bestMatch = candidate
			}
		}
	}

	if bestScore >= 0.6 {
		return bestMatch, bestScore
	}

	return candidates[0], 0.5
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
